src/wrap_algorithms/optimal_fit.rs

```diff
@@ -6,18 +6,107 @@ use crate::wrap_algorithms::WrapAlgorithm;
+/// assigning penalties that work well for monospace text.
+///
+/// If you are wrapping proportional text, you are advised to assign
+/// your own penalties according to your font size. See the individual
+/// penalties below for details.
+///
+/// The underlying wrapping algorithm is implemented by
+/// [`wrap_optimal_fit`], please see that function for examples.
 ///
 /// **Note:** Only available when the `smawk` Cargo feature is
 /// enabled.
-#[derive(Clone, Copy, Debug, Default)]
-pub struct OptimalFit;
+#[derive(Clone, Copy, Debug)]
+pub struct OptimalFit {
+    /// Per-line penalty. This is added for every line, which makes it
+    /// expensive to output more lines than the minimum required.
+    pub nline_penalty: i32,
+
+    /// Per-character cost for lines that overflow the target line width.
+    ///
+    /// With a default value of 50², every single character costs as
+    /// much as leaving a gap of 50 characters behind. This is because
+    /// we assign as cost of `gap * gap` to a short line. When
+    /// wrapping monospace text, we can overflow the line by 1
+    /// character in extreme cases:
+    ///
+    /// ```
+    /// use textwrap::wrap_algorithms::{wrap_optimal_fit, OptimalFit};
+    /// use textwrap::core::Word;
+    ///
+    /// let short = "foo ";
+    /// let long = "x".repeat(50);
+    /// let fragments = vec![Word::from(short), Word::from(&long)];
+    /// let penalties = OptimalFit::new();
+    ///
+    /// // Perfect fit, both words are on a single line with no overflow.
+    /// let wrapped = wrap_optimal_fit(&fragments, &[short.len() + long.len()], &penalties);
+    /// assert_eq!(wrapped, vec![&[Word::from(short), Word::from(&long)]]);
+    ///
+    /// // The words no longer fit, yet we get a single line back. While
+    /// // the cost of overflow (`1 * 2500`) is the same as the cost of the
+    /// // gap (`50 * 50 = 2500`), the tie is broken by `nline_penalty`
+    /// // which makes it cheaper to overflow than to use two lines.
+    /// let wrapped = wrap_optimal_fit(&fragments, &[short.len() + long.len() - 1], &penalties);
+    /// assert_eq!(wrapped, vec![&[Word::from(short), Word::from(&long)]]);
+    ///
+    /// // The cost of overflow would be 2 * 2500, whereas the cost of
+    /// // the gap is only `49 * 49 + nline_penalty = 2401 + 1000 =
+    /// // 3401`. We therefore get two lines.
+    /// let wrapped = wrap_optimal_fit(&fragments, &[short.len() + long.len() - 2], &penalties);
+    /// assert_eq!(wrapped, vec![&[Word::from(short)],
+    ///                          &[Word::from(&long)]]);
+    /// ```
+    ///
+    /// This only happens if the overflowing word is 50 characters
+    /// long _and_ if the word overflows the line by exactly one
+    /// character. If it overflows by more than one character, the
+    /// overflow penalty will quickly outgrow the cost of the gap, as
+    /// seen above.
+    pub overflow_penalty: i32,
+
+    /// When should the last line be considered "short".
```

Comment:
Could you maybe add some more information about the meaning of this value?  I understand that *width* <= *max_width* / *short_line_fraction* is considered a short line, but the description is not entirely clear.

Reply:
Yeah... great point, thanks! I've expanded it a bit now. Does it look better?

An example would of course be great, but I wasn't able to quickly conjure one up with the Wasm demo.

Reply:
Much better, thanks!

Reply:
Thanks, much appreciated!